{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pangenome of colibactin containing Escherichia genomes\n",
    "\n",
    "Construct pan and core genome of 36 colibactin containing Escherichia genomes using Roary software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import BCBio\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from shutil import copyfile\n",
    "import pprint\n",
    "import numpy as np\n",
    "from Bio import Phylo\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_style('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# directories\n",
    "roary_dir = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/'\n",
    "data_dir = '/home/omkar/Projects/panGenome/data/entero_project/genomes/'\n",
    "adj_mat_dir = '/home/omkar/Projects/panGenome/data/entero_project/families/adj_mat/'\n",
    "df_clusters_dir = '/home/omkar/Projects/panGenome/data/entero_project/general/df_clusters.p'\n",
    "df_entero_dir = '/home/omkar/Projects/panGenome/data/entero_project/general/df_entero.p'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subnet_id = 'comp_3'\n",
    "pan_type = 'clusters' # or 'clusters'\n",
    "tgt_genus = 'all'\n",
    "mibig_id = 'BGC0000972.1'\n",
    "mibig_name = 'clb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# main function\n",
    "gff_dir = create_directories(roary_dir, subnet_id, pan_type, tgt_genus, mibig_name)\n",
    "df_clusters, df_entero = get_saved_df(df_clusters_dir, df_entero_dir)\n",
    "fam_tgt_clusters, df_fam_clusters, df_fam_genomes = get_fam_clusters(adj_mat_dir, subnet_id, mibig_id, tgt_genus, df_clusters, df_entero)\n",
    "print('Number of target clusters:', len(fam_tgt_clusters))\n",
    "\n",
    "copy_data_to_roary(data_dir, gff_dir, df_fam_clusters, pan_type)\n",
    "# Run roary \n",
    "# Run on command line \n",
    "# > workon pangenome\n",
    "# > cd /home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/\n",
    "# > roary -f comp_3_clb_Escherichia/ comp_3_clb_Escherichia/gff/*.gff -v -r -e -n -i 90"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_directories(roary_dir, subnet_id, pan_type, tgt_genus, mibig_name):\n",
    "    tgt_dir = os.path.join(roary_dir, pan_type, subnet_id + '_' + mibig_name + '_' + tgt_genus)\n",
    "    if os.path.isdir(tgt_dir):\n",
    "        print(tgt_dir , 'already exists!!')\n",
    "    else:\n",
    "        os.mkdir(tgt_dir)\n",
    "    gff_dir = os.path.join(tgt_dir, 'gff')\n",
    "    if os.path.isdir(gff_dir):\n",
    "        print(gff_dir , 'already exists!!')\n",
    "    else:\n",
    "        os.mkdir(gff_dir)\n",
    "    return gff_dir\n",
    "\n",
    "def get_saved_df(df_clusters_dir, df_entero_dir):\n",
    "    df_clusters = pd.read_pickle(df_clusters_dir)\n",
    "    df_entero = pd.read_pickle(df_entero_dir)\n",
    "    \n",
    "    return df_clusters, df_entero\n",
    "\n",
    "def get_fam_clusters(adj_mat_dir, subnet_id, mibig_id, tgt_genus, df_clusters, df_entero):\n",
    "    df_input_dir = os.path.join(adj_mat_dir, subnet_id)\n",
    "    df_adj_mat = pd.read_csv(df_input_dir)\n",
    "    df_adj_mat = df_adj_mat.set_index(df_adj_mat.columns[0])\n",
    "    df_adj_mat = df_adj_mat.iloc[:,2:-2]\n",
    "    \n",
    "    fam_clusters_list = []\n",
    "    for col in df_adj_mat.columns:\n",
    "        similarity = df_adj_mat.loc[mibig_id,col]\n",
    "        if float(similarity) > 0.3:\n",
    "            fam_clusters_list.append(col)\n",
    "            \n",
    "    idx = pd.IndexSlice\n",
    "\n",
    "    if tgt_genus == 'all':\n",
    "        df_fam_clusters = df_clusters.loc[idx[:,:,:,:,:,:,:,:,:,:,:,fam_clusters_list],:]\n",
    "        fam_patric_ids = df_fam_clusters.index.get_level_values(4).tolist()\n",
    "        fam_tgt_clusters = df_fam_clusters.index.get_level_values(11).tolist()\n",
    "    else:\n",
    "        df_fam_clusters = df_clusters.loc[idx[:,:,:,:,:,:,:,:,:,:,:,fam_clusters_list],:]\n",
    "        df_fam_clusters = df_fam_clusters[df_fam_clusters.index.get_level_values(0) == tgt_genus]\n",
    "        fam_patric_ids = df_fam_clusters.index.get_level_values(4).tolist()\n",
    "        fam_tgt_clusters = df_fam_clusters.index.get_level_values(11).tolist()\n",
    "\n",
    "    df_fam_genomes = df_entero.loc[fam_patric_ids,:]\n",
    "    \n",
    "    return fam_tgt_clusters, df_fam_clusters, df_fam_genomes \n",
    "\n",
    "def copy_data_to_roary(data_dir, gff_dir, df_fam_clusters, pan_type):\n",
    "    fam_patric_ids = df_fam_clusters.index.get_level_values(4).tolist()\n",
    "    fam_clusters_list = df_fam_clusters.index.get_level_values(11).tolist()\n",
    "    \n",
    "    if pan_type == 'genomes':\n",
    "        for pat in fam_patric_ids:\n",
    "            src = os.path.join(data_dir, pat, pat + '_roary.gff')\n",
    "            dest = os.path.join(gff_dir, pat + '.gff')\n",
    "            copyfile(src, dest)\n",
    "            \n",
    "    elif pan_type == 'clusters':\n",
    "        for pat in fam_patric_ids:\n",
    "            src = os.path.join(data_dir, pat, pat + '_roary.gff')\n",
    "            dest = os.path.join(gff_dir, pat + '.gff')\n",
    "            copyfile(src, dest)\n",
    "        cluster_locusID_dict = dict()\n",
    "        \n",
    "        for tmp_id in range(len(fam_patric_ids)):\n",
    "            pat = fam_patric_ids[tmp_id]\n",
    "            clust = fam_clusters_list[tmp_id]\n",
    "            gbk_path = os.path.join(data_dir, pat, 'as4', clust + '.gbk')\n",
    "            cluster_locusID_dict[pat] = []\n",
    "            with open(gbk_path, \"r\") as handle:\n",
    "                for record in SeqIO.parse(handle, \"genbank\"):\n",
    "                    for feature in record.features:\n",
    "                        if feature.type == 'CDS':\n",
    "                            locus_ID = feature.qualifiers['ID'][0]\n",
    "                            cluster_locusID_dict[pat].append(locus_ID)\n",
    "                            \n",
    "        # Curate gff files by adding nucleotide sequence for Roary pangenome\n",
    "        for tmp_id in range(len(fam_patric_ids)):\n",
    "            pat = fam_patric_ids[tmp_id]\n",
    "            clust = fam_clusters_list[tmp_id]\n",
    "            \n",
    "            in_file1 = os.path.join(gff_dir, pat + '.gff')\n",
    "            out_file = os.path.join(gff_dir, clust + '.gff')\n",
    "\n",
    "            with open(in_file1) as in_handle1, open(out_file, 'w') as out_handle:\n",
    "                for line in in_handle1:\n",
    "                    if line.startswith('accn'):\n",
    "                        locus_ID = line.split('\\t')[8].split(';')[0].split('=')[1]\n",
    "                        if locus_ID in cluster_locusID_dict[pat]:\n",
    "                            out_handle.write(line)\n",
    "                    else:\n",
    "                        out_handle.write(line)   \n",
    "        for pat in fam_patric_ids:\n",
    "            in_file1 = os.path.join(gff_dir, pat + '.gff')\n",
    "            if os.path.exists(in_file1):\n",
    "                os.remove(in_file1) \n",
    "                \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colibactin containing clusters : Comp 3 ecoli row BGC0000972.1 \n",
    "adj_mat = pd.read_csv('/home/omkar/Projects/panGenome/data/entero_project/families/adj_mat/comp_3')\n",
    "adj_mat = adj_mat.set_index(adj_mat.columns[0])\n",
    "adj_mat = adj_mat.iloc[:,2:-2]\n",
    "famX_clusters = []\n",
    "for col in adj_mat.columns:\n",
    "    similarity = adj_mat.loc['BGC0000972.1',col]\n",
    "    if float(similarity) > 0.3:\n",
    "        famX_clusters.append(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clusters = pd.read_pickle('/home/omkar/Projects/panGenome/data/entero_project/general/df_clusters.p')\n",
    "# df_ecoli = df_clusters[df_clusters]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pan_ecoli_dir = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_CLB_MIBIG_escherichia/'\n",
    "data_dir = '/home/omkar/Projects/panGenome/data/entero_project/genomes/'\n",
    "os.mkdir(pan_ecoli_dir)\n",
    "# df_ecoli = pd.read_pickle('/home/omkar/Projects/panGenome/data/entero_project/general/df_ecoli.p')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate pangenome of E. coli strains with Colibactin and Yersiniabactin clusters - 30 (28 acc to BiG-SCAPE)\n",
    "pat_clb_ybt_ecoli = df_ecoli.Strain_ID[df_ecoli.PKS_stat == 'Ybt-Clb'].tolist()\n",
    "\n",
    "Ybt_Clb_dir = os.path.join(pan_ecoli_dir, 'Ybt_Clb')\n",
    "os.mkdir(Ybt_Clb_dir)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save genome metadata of Ybt-Clb strains of E coli \n",
    "df_clb = df_ecoli[df_ecoli.PKS_stat == 'Ybt-Clb']\n",
    "df_clb= df_clb.loc[:,['Strain_ID', 'Genome_name', 'Contigs', 'Genome Length', 'GC Content', 'PATRIC CDS', 'RefSeq CDS', 'NCBI Taxon ID','GenBank Accessions', 'RefSeq Accessions', 'Strain', 'Serovar', 'Biovar', 'Pathovar', 'MLST','Publication']]\n",
    "df_clb.set_index('Strain_ID', inplace=True)\n",
    "df_clb.to_csv('/home/omkar/Projects/panGenome/data/entero_project/general/df_clb_ecoli.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_dir = os.path.join(Ybt_Clb_dir, 'gff')\n",
    "os.mkdir(gff_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pat in pat_clb_ybt_ecoli:\n",
    "    src = os.path.join(data_dir, pat, pat + '_roary.gff')\n",
    "    dest = os.path.join(gff_dir, pat + '.gff')\n",
    "    copyfile(src, dest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run on command line \n",
    "# > workon pangenome\n",
    "# > cd /home/omkar/Projects/panGenome/data/entero_project/pan_roary/ecoli/\n",
    "# > roary -f Ybt_Clb/ Ybt_Clb/gff/*.gff -v -r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# directories\n",
    "roary_dir = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/'\n",
    "data_dir = '/home/omkar/Projects/panGenome/data/entero_project/genomes/'\n",
    "df_clusters_dir = '/home/omkar/Projects/panGenome/data/entero_project/general/df_clusters.p'\n",
    "df_entero_dir = '/home/omkar/Projects/panGenome/data/entero_project/general/df_entero.p'\n",
    "homology_dir = '/home/omkar/Projects/panGenome/data/entero_project/homology/'\n",
    "# variables \n",
    "subnet_id = 'comp_3'\n",
    "pan_type = 'clusters' # or 'clusters'\n",
    "tgt_genus = 'all'\n",
    "mibig_id = 'BGC0000972.1'\n",
    "mibig_name = 'clb'\n",
    "core_cutoff = 1\n",
    "\n",
    "# main function\n",
    "roary_out = get_roary_out(roary_dir, pan_type, subnet_id, mibig_name, tgt_genus)\n",
    "df_roary, df_roary_info, df_clusters, df_entero = get_saved_df(df_clusters_dir, df_entero_dir, roary_out)\n",
    "df_roary_filtered, df_roary_bool, df_roary_sorted = filter_roary_pangenome(df_roary, core_cutoff)\n",
    "\n",
    "get_pan_freq_plot(df_roary_sorted, roary_out)\n",
    "get_pan_pie_chart(df_roary_sorted, roary_out)\n",
    "get_core_pan_curve(roary_out)\n",
    "get_presence_heatmap(df_roary_filtered, roary_out, df_entero)\n",
    "\n",
    "# get_ecoli_genomes(df_entero)\n",
    "\n",
    "# df_roary_filtered.columns = [df_entero.loc[pat_id, 'Strain'] for pat_id in df_roary_filtered.columns]\n",
    "\n",
    "# sns.clustermap(df_roary_filtered, row_cluster=True, cmap = plt.cm.Blues, yticklabels = False, figsize=(15,20))\n",
    "# df_entero_pks = df_entero.loc[df_roary.columns,:]\n",
    "\n",
    "# nissle_genomes = ['316435.10', '316435.29']\n",
    "# upec_genomes = df_entero_pks[df_entero_pks.Pathovar == 'UPEC'].index.tolist()\n",
    "# expec_genomes = df_entero_pks[df_entero_pks.Pathovar == 'ExPEC'].index.tolist()\n",
    "# aiec_genomes = df_entero_pks[df_entero_pks.Pathovar == 'AIEC'].index.tolist()\n",
    "\n",
    "# df_nissle_absent = get_nissle_absent_genes(df_roary_bool, nissle_genomes, upec_genomes, expec_genomes, aiec_genomes)\n",
    "# df_nissle_absent.columns = [df_entero.loc[pat_id, 'Strain'] for pat_id in df_nissle_absent.columns]\n",
    "# nissle_absent_gene_list = df_nissle_absent.index.get_level_values('Gene')\n",
    "# sns.set(font_scale=1.6)\n",
    "# sns.clustermap(df_nissle_absent, cmap = plt.cm.Blues, yticklabels = nissle_absent_gene_list, figsize=(15,20))\n",
    "\n",
    "# df_nissle_absent_features = get_nissle_absent_features(df_roary_info, '199310.4', df_nissle_absent, data_dir)\n",
    "# homology_out_dir = create_homology_dir(homology_dir, mibig_name, tgt_genus)\n",
    "\n",
    "# df_pan_features = pd.read_pickle(os.path.join(homology_out_dir, 'df_pan_features_unflitered.p'))\n",
    "# df_interpro_all = pd.read\n",
    "\n",
    "# run interpro on entire pangenome\n",
    "# sudo ./interproscan.sh -i /home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_clb_Escherichia/_1553187017/pan_genome_reference.fa -b /home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_clb_Escherichia/inteproscan/ -goterms -pa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_roary_filtered = df_roary_sorted[df_roary_sorted.sum(1)>9]\n",
    "# df_roary_sorted.sum(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "df_clusters_clb = df_clusters[df_clusters.index.get_level_values(-1).isin(df_roary_filtered.columns)]\n",
    "genus_list = df_clusters_clb.index.get_level_values(0)\n",
    "\n",
    "pickle_out = open('/home/omkar/Projects/panGenome/data/entero_project/general/genus_colors.p',\"rb\")\n",
    "my_palette = pickle.load(pickle_out)\n",
    "pickle_out.close()\n",
    "col_colors = genus_list.map(my_palette)\n",
    "\n",
    "ax = sns.clustermap(df_roary_filtered, row_cluster=True, yticklabels = df_roary_filtered.index.get_level_values(0), col_cluster=True, cmap = 'Greens',figsize=(40,80), col_colors=col_colors)\n",
    "# # ax2.cax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax2 = sns.clustermap(ax.data2d.iloc[:,:36], row_cluster=True, col_cluster=True, cmap = 'Greens',figsize=(40,80))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# col_list = df_roary_bool.columns.tolist() \n",
    "# col_list[5] = 'Nissle 1917_2'\n",
    "\n",
    "# df_roary_bool.columns = col_list\n",
    "\n",
    "df_roary_block1 = df_roary_bool[df_roary_bool.sum(1)>=0.85*df_roary_bool.shape[1]]\n",
    "\n",
    "df_roary_block2 = df_roary_bool[(df_roary_bool.sum(axis=1) < 0.85 * df_roary_bool.shape[1]) & (df_roary_bool.sum(axis=1) > 0.35*df_roary_bool.shape[1])]\n",
    "ax2 = sns.clustermap(df_roary_block2, row_cluster=True, col_cluster=True, cmap = 'Greens', yticklabels = False, figsize=(13,21))\n",
    "ax2.cax.set_visible(False)\n",
    "\n",
    "roary_all_index = df_roary_block1.index.get_level_values(0).tolist() + ax2.data2d.index.get_level_values(0).tolist()\n",
    "idx = pd.IndexSlice\n",
    "df_roary_all = df_roary_bool.loc[idx[roary_all_index,:], ax2.data2d.columns]\n",
    "cmap = [sns.color_palette('Greens')[0] , sns.color_palette('Greens')[-1]]\n",
    "ax_all = sns.clustermap(df_roary_all, row_cluster=False, col_cluster=False, cmap = cmap, yticklabels = False, figsize=(16,21))\n",
    "ax_all.cax.set_visible(False)\n",
    "ax_all.ax_heatmap.set_ylabel('')\n",
    "plt.savefig(os.path.join(roary_out, 'fig2_presence_heatmap_all.png'),pad_inches = 0.1, bbox_inches = 'tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_core_pan_curve(roary_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_roary_block2 = df_roary_bool[(df_roary_bool.sum(axis=1) < df_roary_bool.shape[1]) & (df_roary_bool.sum(axis=1) > 5)]\n",
    "ax2 = sns.clustermap(df_roary_block2, row_cluster=True, col_cluster=True, cmap = 'Greens', yticklabels = True, figsize=(13,21))\n",
    "ax2.cax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_path = os.path.join(roary_out,'number_of_conserved_genes.Rtab')\n",
    "pan_path = os.path.join(roary_out,'number_of_genes_in_pan_genome.Rtab')\n",
    "    \n",
    "df_core_cnt = pd.read_table(core_path, sep='\\t')\n",
    "df_pan_cnt = pd.read_table(pan_path, sep='\\t')\n",
    "df_pan_cnt\n",
    "\n",
    "print('absense: ',[255*col for col in sns.color_palette('Greens')[0]])\n",
    "print('presense: ',[255*col for col in sns.color_palette('Greens')[-1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import fasta file with pangenome reference\n",
    "pan_ref_path = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_CLB_MIBIG/_1553084326/pan_genome_reference.fa'\n",
    "pan_ref_filtered_path = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_CLB_MIBIG/pan_genome_filtered.fa'\n",
    "pan_ref_filtered_trna_path = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_CLB_MIBIG/pan_genome_filtered.frn'\n",
    "\n",
    "\n",
    "pan_filtered_genes = roary_filtered.index.get_level_values('Gene').tolist()\n",
    "pat_strains_dict = dict()\n",
    "gene_strains_dict = dict()\n",
    "\n",
    "for record in SeqIO.parse(pan_ref_path, \"fasta\"):\n",
    "    id_list = record.description.split(' ')\n",
    "    gene = id_list[1]\n",
    "    if gene in pan_filtered_genes:\n",
    "        pat_gene_id = id_list[0]\n",
    "        if 'peg' in pat_gene_id:\n",
    "            tmp_list = pat_gene_id.split('.peg')\n",
    "        elif  'rna' in pat_gene_id:\n",
    "            tmp_list = pat_gene_id.split('.rna')\n",
    "        else:\n",
    "            print('Errror: Extra gene id :', pat_gene_id)\n",
    "        pat_strain_id = tmp_list[0][4:]\n",
    "\n",
    "        if pat_strain_id not in pat_strains_dict.keys():\n",
    "            pat_strains_dict[pat_strain_id] = [pat_gene_id]\n",
    "            gene_strains_dict[pat_gene_id] = gene\n",
    "        else:\n",
    "            pat_strains_dict[pat_strain_id].append(pat_gene_id)\n",
    "            gene_strains_dict[pat_gene_id] = gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/home/omkar/Projects/panGenome/data/entero_project/genomes/'\n",
    "# Get protein fasta sequences for these genes to generate pangenome for homology analysis \n",
    "record_list = []\n",
    "trna_list = []\n",
    "\n",
    "for pat_strain_id in pat_strains_dict.keys():\n",
    "    faa_path = os.path.join(data_dir, pat_strain_id, pat_strain_id + '.PATRIC.faa')\n",
    "    rna_path = os.path.join(data_dir, pat_strain_id, pat_strain_id + '.PATRIC.frn')\n",
    "\n",
    "    for record in SeqIO.parse(faa_path, \"fasta\"):\n",
    "        id_list = record.id.split('|')\n",
    "        pat_gene_id = 'fig|' + id_list[1]\n",
    "        print(pat_gene_id)\n",
    "        if pat_gene_id in pat_strains_dict[pat_strain_id]:\n",
    "            record.description = record.description + ', ' + gene_strains_dict[pat_gene_id]\n",
    "            record_list.append(record)\n",
    "            \n",
    "            \n",
    "    for record in SeqIO.parse(rna_path, \"fasta\"):\n",
    "        id_list = record.id.split('|')\n",
    "        pat_gene_id = 'fig|' + id_list[1]\n",
    "        print(pat_gene_id)\n",
    "        if pat_gene_id in pat_strains_dict[pat_strain_id]:\n",
    "            record.description = record.description + ', ' + gene_strains_dict[pat_gene_id]\n",
    "            trna_list.append(record)  \n",
    "            \n",
    "SeqIO.write(record_list, pan_ref_filtered_path, \"fasta\")\n",
    "SeqIO.write(trna_list, pan_ref_filtered_trna_path, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Write gene features file for pangenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load roary file again\n",
    "Ybt_Clb_path = '/home/omkar/Projects/panGenome/data/entero_project/pan_roary/genomes/comp_3_CLB_MIBIG/_1553084326/gene_presence_absence.csv'\n",
    "roary = pd.read_table(Ybt_Clb_path,\n",
    "                     sep=',',\n",
    "                     low_memory=False)\n",
    "roary.set_index(['Gene', 'Annotation'], inplace=True)\n",
    "\n",
    "# Drop the other info columns\n",
    "roary.drop(list(roary.columns[:12]), axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get features dataframes\n",
    "# Only for CDS at the moment, include RNAs later\n",
    "\n",
    "CFT_features_path = os.path.join(data_dir, '199310.4', '199310.4.PATRIC.features.tab') \n",
    "df_CFT_features = pd.read_csv(CFT_features_path, sep='\\t')\n",
    "df_CFT_features = df_CFT_features[df_CFT_features.feature_type == 'CDS']\n",
    "df_CFT_features = df_CFT_features.loc[:,['genome_name', 'patric_id', 'refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', 'gene', 'product', 'figfam_id',\n",
    "       'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway']]\n",
    "df_CFT_features = df_CFT_features.set_index('patric_id')\n",
    "\n",
    "IHE_features_path = os.path.join(data_dir, '714962.3', '714962.3.PATRIC.features.tab') \n",
    "df_IHE_features = pd.read_csv(IHE_features_path, sep='\\t')\n",
    "df_IHE_features = df_IHE_features[df_IHE_features.feature_type == 'CDS']\n",
    "df_IHE_features = df_IHE_features.loc[:,['genome_name', 'patric_id', 'refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', 'gene', 'product', 'figfam_id',\n",
    "       'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway']]\n",
    "df_IHE_features = df_IHE_features.set_index('patric_id')\n",
    "\n",
    "Nissle_features_path = os.path.join(data_dir, '316435.10', '316435.10.PATRIC.features.tab') \n",
    "df_Nissle_features = pd.read_csv(Nissle_features_path, sep='\\t')\n",
    "df_Nissle_features = df_Nissle_features[df_Nissle_features.feature_type == 'CDS']\n",
    "df_Nissle_features = df_Nissle_features.loc[:,['genome_name', 'patric_id', 'refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', 'gene', 'product', 'figfam_id',\n",
    "       'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway']]\n",
    "df_Nissle_features = df_Nissle_features.set_index('patric_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import fasta file with pangenome reference\n",
    "data_dir = '/home/omkar/Projects/panGenome/data/entero_project/genomes/'\n",
    "\n",
    "# Get .PATRIC.features for these genes to generate features of pangenome for homology analysis \n",
    "# Priority given to CFT073, IHE3034, Nissle 1917 and then whatever is chosen in pan_fasta file features\n",
    "df_pan_features = pd.DataFrame(columns=['genome_name', 'pan_gene_id','refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', 'gene', 'product', 'figfam_id',\n",
    "       'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway']) \n",
    "\n",
    "pan_filtered_genes = roary.index.get_level_values('Gene').tolist()\n",
    "pat_strains_dict = dict()\n",
    "gene_strains_dict = dict()\n",
    "\n",
    "\n",
    "for record in SeqIO.parse(pan_ref_path, \"fasta\"):\n",
    "    id_list = record.description.split(' ')\n",
    "    gene = id_list[1]\n",
    "    pat_gene_id = id_list[0]\n",
    "        \n",
    "    if gene in pan_filtered_genes:\n",
    "        if not roary.loc[gene,'199310.4'].isnull().any().any():\n",
    "            CFT_gene_id = roary.loc[gene,'199310.4'].values[0]\n",
    "            \n",
    "            if CFT_gene_id in df_CFT_features.index:\n",
    "                df_pan_features.loc[CFT_gene_id,:] = df_CFT_features.loc[CFT_gene_id, :]\n",
    "                df_pan_features.loc[CFT_gene_id,'pan_gene_id'] = gene\n",
    "                df_pan_features.loc[CFT_gene_id,'pan_pat_id'] = pat_gene_id\n",
    "            else:\n",
    "                print('1', gene)\n",
    "                \n",
    "        elif not roary.loc[gene,'714962.3'].isnull().any().any():\n",
    "            IHE_gene_id = roary.loc[gene,'714962.3'].values[0]\n",
    "            \n",
    "            if IHE_gene_id in df_IHE_features.index:\n",
    "                df_pan_features.loc[IHE_gene_id,:] = df_IHE_features.loc[IHE_gene_id, :]\n",
    "                df_pan_features.loc[IHE_gene_id,'pan_gene_id'] = gene\n",
    "                df_pan_features.loc[IHE_gene_id,'pan_pat_id'] = pat_gene_id\n",
    "            else:\n",
    "                print('2', gene)            \n",
    "            \n",
    "        elif not roary.loc[gene,'316435.10'].isnull().any().any():\n",
    "            Nissle_gene_id = roary.loc[gene,'316435.10'].values[0]\n",
    "            \n",
    "            if Nissle_gene_id in df_Nissle_features.index:\n",
    "                df_pan_features.loc[Nissle_gene_id,:] = df_Nissle_features.loc[Nissle_gene_id, :]\n",
    "                df_pan_features.loc[Nissle_gene_id,'pan_gene_id'] = gene\n",
    "                df_pan_features.loc[Nissle_gene_id,'pan_pat_id'] = pat_gene_id\n",
    "            else:\n",
    "                print('4', gene)\n",
    "            \n",
    "                \n",
    "        else:\n",
    "            if 'peg' in pat_gene_id:\n",
    "                tmp_list = pat_gene_id.split('.peg')\n",
    "                pat_strain_id = tmp_list[0][4:]\n",
    "                \n",
    "                other_features_path = os.path.join(data_dir, pat_strain_id, pat_strain_id + '.PATRIC.features.tab') \n",
    "                \n",
    "                df_other_features = pd.read_csv(other_features_path, sep='\\t')\n",
    "                df_other_features = df_other_features[df_other_features.feature_type == 'CDS']\n",
    "                df_other_features = df_other_features.loc[:,['patric_id', 'refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', \n",
    "                                'gene', 'product', 'figfam_id', 'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway']]\n",
    "                df_other_features = df_other_features.set_index('patric_id')\n",
    "                \n",
    "                if pat_gene_id in df_other_features.index:\n",
    "                    df_pan_features.loc[pat_gene_id,:] = df_other_features.loc[pat_gene_id, :]\n",
    "                    df_pan_features.loc[pat_gene_id,'pan_gene_id'] = gene\n",
    "                    df_pan_features.loc[pat_gene_id,'pan_pat_id'] = pat_gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bigscape_dir = '/home/omkar/Projects/panGenome/data/entero_project/bigscape_specific_inputs/comp_3_clb_all/'\n",
    "clusters_tmp_list = os.listdir(bigscape_dir)\n",
    "for cluster_id in clusters_tmp_list:\n",
    "    if cluster_id[:-4] not in fam_tgt_clusters:\n",
    "        print(cluster_id)\n",
    "    else:\n",
    "        print('ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pan_features.to_pickle('/home/omkar/Projects/panGenome/data/entero_project/homology/pan_clb_cluster/homology_maps/pan_features_df.p')\n",
    "df_pan_features.to_csv('/home/omkar/Projects/panGenome/data/entero_project/homology/pan_clb_cluster/homology_maps/pan_features_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "import os\n",
    "import BCBio\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from shutil import copyfile\n",
    "import pprint\n",
    "import numpy as np\n",
    "from Bio import Phylo\n",
    "from scipy.cluster.hierarchy import linkage\n",
    "\n",
    "def get_core_pan_curve(roary_out):\n",
    "    core_path = os.path.join(roary_out,'number_of_conserved_genes.Rtab')\n",
    "    pan_path = os.path.join(roary_out,'number_of_genes_in_pan_genome.Rtab')\n",
    "    \n",
    "    df_core_cnt = pd.read_table(core_path, sep='\\t')\n",
    "    df_pan_cnt = pd.read_table(pan_path, sep='\\t')\n",
    "    \n",
    "    avg_core = df_core_cnt.sum(0)/df_core_cnt.shape[0]\n",
    "    max_core = df_core_cnt.max(0)\n",
    "    min_core = df_core_cnt.min(0)\n",
    "    \n",
    "    avg_pan = df_pan_cnt.sum(0)/df_core_cnt.shape[0]\n",
    "    max_pan = df_pan_cnt.max(0)\n",
    "    min_pan = df_pan_cnt.min(0)\n",
    "    \n",
    "    genome_numbers = list(range(1,df_core_cnt.shape[1]+1))\n",
    "    \n",
    "    plt.figure(figsize=(8,8))\n",
    "    \n",
    "    plt.plot(genome_numbers, avg_core, color='green', linewidth=4)\n",
    "    plt.plot(genome_numbers, avg_pan, color='lightgreen', linewidth=4)\n",
    "    \n",
    "    \n",
    "    plt.plot(genome_numbers, min_core, color='green', linestyle='dashed', linewidth=4)\n",
    "    plt.plot(genome_numbers, max_core, color='green', linestyle='dashed', linewidth=4)\n",
    "    \n",
    "    plt.plot(genome_numbers, min_pan,  color='lightgreen',linestyle='dashed', linewidth=4)\n",
    "    plt.plot(genome_numbers, max_pan, color='lightgreen', linestyle='dashed', linewidth=4)\n",
    "    \n",
    "    plt.fill_between(genome_numbers, avg_core, y2=0, color = sns.color_palette('Greens')[3])\n",
    "    plt.fill_between(genome_numbers, avg_pan, y2=avg_core, color = sns.color_palette('Greens')[1])\n",
    "    \n",
    "    plt.ylim(ymin=0)\n",
    "    plt.xlim(xmin=1,xmax=36)\n",
    "    plt.xlabel('No. of genomes')\n",
    "    plt.ylabel('No. of gene families')\n",
    "    \n",
    "#     legend([p2, p1], [\"line 2\", \"line 1\"])\n",
    "    \n",
    "    plt.legend(['Core genome', 'Pan genome'])\n",
    "    \n",
    "    plt.savefig(os.path.join(roary_out, 'fig3_pan_core_curves.png'))\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "def get_nissle_absent_features(df_roary_info, CFT_pat, df_nissle_absent, data_dir):\n",
    "    CFT_gene_id_list = df_roary_info.loc[df_nissle_absent.index, CFT_pat].tolist()\n",
    "    \n",
    "    CFT_features_path = os.path.join(data_dir, CFT_pat, CFT_pat + '.PATRIC.features.tab') \n",
    "    df_CFT_features = pd.read_csv(CFT_features_path, sep='\\t')\n",
    "    df_CFT_features = df_CFT_features[df_CFT_features.feature_type == 'CDS']\n",
    "    df_CFT_features = df_CFT_features.loc[:,['genome_name', 'patric_id', 'refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', 'gene', 'product', 'figfam_id',\n",
    "       'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway', 'CFT_gene_id']]\n",
    "    df_CFT_features = df_CFT_features.set_index('patric_id')\n",
    "       \n",
    "        \n",
    "    df_nissle_absent_features = pd.DataFrame(columns=['genome_name', 'refseq_locus_tag', 'alt_locus_tag', 'uniprotkb_accession', 'gene', 'product', 'figfam_id',\n",
    "       'plfam_id', 'pgfam_id', 'go', 'ec', 'pathway']) \n",
    "    temp_cnt = 0\n",
    "    for CFT_gene_id in CFT_gene_id_list:\n",
    "        pan_gene_id = df_nissle_absent.index.get_level_values('Gene')[temp_cnt]\n",
    "        if CFT_gene_id in df_CFT_features.index:\n",
    "            df_nissle_absent_features.loc[pan_gene_id,'CFT_gene_id'] = CFT_gene_id\n",
    "            df_nissle_absent_features.loc[pan_gene_id,:] = df_CFT_features.loc[CFT_gene_id, :]\n",
    "        temp_cnt = temp_cnt + 1    \n",
    "            \n",
    "    return df_nissle_absent_features\n",
    "\n",
    "def get_nissle_absent_genes(df_roary_bool, nissle_genomes, upec_genomes, expec_genomes, aiec_genomes):\n",
    "    df_nissle_absent = df_roary_bool[(df_roary_bool.sum(axis=1) < df_roary_bool.shape[1]) & \n",
    "                         (df_roary_bool[nissle_genomes[0]] == 0) & \n",
    "                         (df_roary_bool[nissle_genomes[1]] == 0) & \n",
    "                         (df_roary_bool[upec_genomes[0]] == 1) & \n",
    "                         (df_roary_bool[upec_genomes[1]] == 1) & \n",
    "                         (df_roary_bool[upec_genomes[2]] == 1) & \n",
    "                         (df_roary_bool[upec_genomes[3]] == 1) & \n",
    "                         (df_roary_bool[upec_genomes[4]] == 1) & \n",
    "                         (df_roary_bool[upec_genomes[5]] == 1) & \n",
    "                         (df_roary_bool[expec_genomes[0]] == 1) &\n",
    "                         (df_roary_bool[aiec_genomes[0]] == 1)]\n",
    "    \n",
    "    df_nissle_absent = df_nissle_absent.drop(columns=nissle_genomes)\n",
    "    \n",
    "    return df_nissle_absent\n",
    "    \n",
    "def create_homology_dir(homology_dir, mibig_name, tgt_genus):\n",
    "    tmp_dir = os.path.join(homology_dir, 'pan_' + mibig_name + '_vs_' + tgt_genus)\n",
    "    if os.path.isdir(tmp_dir):\n",
    "        print(tmp_dir, 'already exists!')\n",
    "    else:\n",
    "        os.mkdir(tmp_dir)\n",
    "        \n",
    "    tgt_dir = os.path.join(tmp_dir, 'tgt_faa')\n",
    "    if os.path.isdir(tgt_dir):\n",
    "        print(tgt_dir, 'already exists!')\n",
    "    else:\n",
    "        os.mkdir(tgt_dir)\n",
    "        \n",
    "    temp_dir = os.path.join(tmp_dir, 'temp_faa')\n",
    "    if os.path.isdir(temp_dir):\n",
    "        print(tgt_dir, 'already exists!')\n",
    "    else:\n",
    "        os.mkdir(temp_dir)\n",
    "    return tmp_dir\n",
    "\n",
    "def filter_roary_pangenome(df_roary, core_cutoff):\n",
    "    df_roary_bool = df_roary.replace('.{2,100}', 1, regex=True)\n",
    "    df_roary_bool.replace(np.nan, 0, regex=True, inplace=True)    \n",
    "    df_roary_filtered = df_roary_bool[(df_roary_bool.sum(axis=1) < df_roary_bool.shape[1]) & (df_roary_bool.sum(axis=1) > 1)]\n",
    "    \n",
    "    # Sort the matrix by the sum of strains presence\n",
    "    idx = df_roary_bool.sum(axis=1).sort_values(ascending=False).index\n",
    "    df_roary_sorted = df_roary_bool.ix[idx]\n",
    "    \n",
    "    return df_roary_filtered, df_roary_bool, df_roary_sorted\n",
    "\n",
    "\n",
    "def get_roary_out(roary_dir, pan_type, subnet_id, mibig_name, tgt_genus):\n",
    "    roary_tmp = os.path.join(roary_dir, pan_type, subnet_id + '_' + mibig_name + '_' + tgt_genus)\n",
    "    file_list = [file for file in os.listdir(roary_tmp) if '_' in file]\n",
    "    print(file_list)\n",
    "    if len(file_list) > 1:\n",
    "        print('Too many output pangenomes!!')\n",
    "        return None\n",
    "    else:\n",
    "        roary_out = os.path.join(roary_tmp, file_list[0])\n",
    "    return roary_out\n",
    "\n",
    "def get_saved_df(df_clusters_dir, df_entero_dir, roary_out):\n",
    "    df_clusters = pd.read_pickle(df_clusters_dir)\n",
    "    df_entero = pd.read_pickle(df_entero_dir)\n",
    "    \n",
    "    gene_presence_path = os.path.join(roary_out, 'gene_presence_absence.csv')\n",
    "    df_roary = pd.read_table(gene_presence_path, sep=',', low_memory=False)\n",
    "    # Set index (group name)\n",
    "    df_roary.set_index(['Gene', 'Annotation'], inplace=True)\n",
    "    df_roary_info = df_roary.copy(deep = True)\n",
    "    \n",
    "    # Drop the other info columns\n",
    "    df_roary = df_roary.drop(list(df_roary.columns[:12]), axis=1)\n",
    "    return df_roary, df_roary_info, df_clusters, df_entero\n",
    "\n",
    "def get_pan_freq_plot(roary, roary_out):\n",
    "    plt.figure(figsize=(10, 7))\n",
    "    plt.hist(roary.sum(axis=1), roary.shape[1], color='purple',\n",
    "         histtype=\"stepfilled\", alpha=.7)\n",
    "    plt.xlabel('Number of genomes', fontweight = 'bold', fontsize = '20')\n",
    "    plt.ylabel('Number of genes', fontweight = 'bold', fontsize = '20')\n",
    "    plt.xticks(fontsize = '18')\n",
    "    plt.yticks(fontsize = '18')\n",
    "    \n",
    "    plt.savefig(os.path.join(roary_out, 'fig1_pan_freq.png'),pad_inches = 0.1, bbox_inches = 'tight')\n",
    "    \n",
    "def get_pan_pie_chart(roary, roary_out):\n",
    "    # Plot the pangenome pie chart\n",
    "    plt.figure(figsize=(10, 10))\n",
    "\n",
    "    core = roary[roary.sum(axis=1) == roary.shape[1]].shape[0]\n",
    "    softcore = roary[(roary.sum(axis=1) < roary.shape[1]) &\n",
    "                     (roary.sum(axis=1) >= roary.shape[1]*0.85)].shape[0]\n",
    "    shell = roary[(roary.sum(axis=1) < roary.shape[1]*0.85) &\n",
    "                     (roary.sum(axis=1) >= roary.shape[1]*0.15)].shape[0]\n",
    "    cloud = roary[roary.sum(axis=1) < roary.shape[1]*0.15].shape[0]\n",
    "\n",
    "    total = roary.shape[0]\n",
    "    \n",
    "    def my_autopct(pct):\n",
    "        val=int(pct*total/100.0)\n",
    "        return '{v:d}'.format(v=val)\n",
    "    a=plt.pie([core, softcore, shell, cloud],\n",
    "      labels=['core  (%d strains)'%roary.shape[1],\n",
    "              'soft-core  (%d <= strains < %d)'%(roary.shape[1]*.85,\n",
    "                                                 roary.shape[1]),\n",
    "              'shell\\n(%d <= strains < %d)'%(roary.shape[1]*.15,\n",
    "                                             roary.shape[1]*.85),\n",
    "              'cloud\\n(strains < %d)'%(roary.shape[1]*.15)],\n",
    "      explode=[0.1, 0.05, 0.02, 0], radius=0.9,\n",
    "      colors=[(0, 0, 1, float(x)/total) for x in (core, softcore, shell, cloud)],\n",
    "      autopct=my_autopct, textprops={'fontsize': 20, 'fontweight': 'bold'})\n",
    "    \n",
    "    plt.savefig(os.path.join(roary_out, 'fig2_pan_pie.png'),pad_inches = 0.1, bbox_inches = 'tight')\n",
    "    \n",
    "def get_presence_heatmap(roary, roary_out, df_entero):\n",
    "#     roary.columns = [df_entero.loc[pat_id, 'Strain'] for pat_id in roary.columns]\n",
    "    ax = sns.clustermap(roary, row_cluster=True, col_cluster=True, cmap = 'Greens', yticklabels = False, figsize=(13,21))\n",
    "    ax.cax.set_visible(False)\n",
    "    \n",
    "    plt.savefig(os.path.join(roary_out, 'fig2_presence_heatmap.png'),pad_inches = 0.1, bbox_inches = 'tight')\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
